src/module.c

```diff
@@ -3048,6 +3051,264 @@ int RM_HashGet(RedisModuleKey *key, int flags, ...) {
+ *   returned. Can be NULL if `AUTOID` is set, if you don't care to receive the
+ *   ID. If `AUTOID` is not set, this is the requested ID.
+ * - `argv`: A pointer to an array of size `numfields * 2` containing the
+ *   fields and values.
+ * - `numfields`: The number of field-value pairs in `argv`.
+ *
+ * Returns REDISMODULE_OK if an entry has been added. Returns REDISMODULE_ERR if
+ * the key is not opened for writing, if the key has a type other than stream,
+ * if the key doesn't exist in case the flag `NOMKSTREAM` is set, if unknown
+ * flags are given or when attempting to create an entry with an id which is not
+ * allowed.
+ */
+int RM_StreamAdd(RedisModuleKey *key, int flags, RedisModuleStreamID *id, RedisModuleString **argv, long numfields) {
+    /* Validate args */
+    if (flags & ~(REDISMODULE_STREAM_AUTOID |
+                  REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR; /* unknown flags */
+    if (!(key->mode & REDISMODULE_WRITE))
+        return REDISMODULE_ERR;
+    if (key->value && key->value->type != OBJ_STREAM)
+        return REDISMODULE_ERR;
+    if (key->value == NULL && (flags & REDISMODULE_STREAM_NOMKSTREAM))
+        return REDISMODULE_ERR;
+    if (!(flags & REDISMODULE_STREAM_AUTOID) &&
+        (id == NULL || (id->ms == 0 && id->seq == 0)))
+        return REDISMODULE_ERR;
+
+    /* Create key if necessery */
+    if (key->value == NULL) {
+        moduleCreateEmptyKey(key, REDISMODULE_KEYTYPE_STREAM);
+    }
+
+    stream *s = key->value->ptr;
+    streamID added_id;
+    streamID use_id;
+    streamID *use_id_ptr = NULL;
+    if (!(flags & REDISMODULE_STREAM_AUTOID)) {
+        use_id.ms = id->ms;
+        use_id.seq = id->seq;
+        use_id_ptr = &use_id;
+    }
+    if (streamAppendItem(s, argv, numfields, &added_id, use_id_ptr) == C_ERR) {
+        /* Current last ID is greater than or equal to 'use_id' */
+        return REDISMODULE_ERR;
+    }
+
+    if (id != NULL) {
+        id->ms = added_id.ms;
+        id->seq = added_id.seq;
+    }
+
+    return REDISMODULE_OK;
+}
+
+/* Sets up a stream iterator.
+ *
+ * - `key`: The stream key opened for reading using RedisModule_OpenKey().
+ * - `flags`:
+ *   - `REDISMODULE_STREAM_EXCLUSIVE`: Don't include minid and maxid in the
+ *     iterated range.
+ *   - `REDISMODULE_STREAM_REVERSE`: Iterate in reverse order, starting from
+ *     maxid.
+ * - `minid`: Lower bound of the iterated range. May be REDISMODULE_STREAM_MIN.
+ * - `maxid`: Upper bound of the iterated range. May be REDISMODULE_STREAM_MAX.
+ *
+ * Returns REDISMODULE_OK on success and REDISMODULE_ERR if the key doesn't
+ * refer to a stream or if invalid arguments were given.
+ *
+ * The stream items are retrieved using RedisModule_StreamIteratorNext(). The
+ * iterator must be freed by callling RedisModule_StreamIteratorStop().
+ */
+int RM_StreamIteratorStart(RedisModuleKey *key, int flags, RedisModuleStreamID *minid, RedisModuleStreamID *maxid) {
```

Comment:
maybe call them `start` and `end`?
it's a bit weird that in case of REDISMODULE_STREAM_REVERSE minid > maxid

also to avoid confusion with XTRIM MINID

Reply:
If REVERSE is specified, maxid is actually the start and minid is the end, so I thought minid and maxid are more clear than startid and endid. Maybe rename to upper and lower? Or I can change the semantics and require startid >= endid if REVERSE is specified. Or maybe just do reverse implicitly if startid > endid and we don't need a REVERSE flag...?

Reply:
i think it's best to stick with the standard set by `streamIteratorStart` and use `start` and `end`
(and then REVERSE goes from `end` to `start` which kind makes sense..?)

Reply:
One more thing: If startid and endid are **not** pointers, REDISMODULE_STREAM_MIN and REDISMODULE_STREAM_MAX can be used directly. OK?

Reply:
why do you want to pass them by value? i think it's more intuitive to pass NULL the means "i want the edge"

Reply:
also please you called them REDISMODULE_STREAMID_MIN and not REDISMODULE_STREAM_MIN (maybe update comment)

Reply:
OK, so let's use pointers and NULL for the edges. Then, we don't really need MIN and MAX, so I'll delete them. Better just have one way of specifying it, right?

(I do think start and end are ambiguous. Start and end could mean (1) where you start and end the iteration vs (2) the start and end of the interval. But I'll stick with the start-end terminology and make it clear in the docs.)